### PR TITLE
Charming actions workflows templates

### PR DESCRIPTION
**Pull request by natalian98**
The aim of this PR is to find a common agreement between charming teams on how to set up the [charming actions workflows](https://github.com/canonical/charming-actions) for publishing (releasing to edge) and releasing to other tracks and channels.

These workflows concern single charm repos, for multiple ones `charm-name` is additionally required as input and `charm-path` needs to be added to the job.

Please comment on this proposal so that we can discuss and figure out 'official' templates.

### CODE REVIEW

File: .github/workflows/release.yaml
Comment (natalian98) [inline .github/workflows/release.yaml:11]: According to the release template, origin channel is required for single charm repos, but not for multi ones. @agathanatasha do you know if there a reason behind this difference?

File: .github/workflows/publish.yaml
Comment (simskij) [inline .github/workflows/publish.yaml:20]: we **could** have this check, but on the other hand, we could drop it as the `branch/` part now is optional to release to charmhub branches.

https://github.com/canonical/charming-actions/blob/b01573cbc75f365ff02b5925ffc2d21f76246d93/src/services/ref/ref.ts#L44-L63

Comment (natalian98) [reply .github/workflows/publish.yaml:20]: Thanks @simskij. Do you think we should include some checks that are present e.g. [in prometheus-k8s charm](https://github.com/canonical/prometheus-k8s-operator/blob/main/.github/workflows/release-edge.yaml#L109)?

Comment (simskij) [reply .github/workflows/publish.yaml:20]: That would make sense. We usually don't want to release something that we know is broken 👍🏼

File: .github/workflows/release.yaml
Comment (agathanatasha) [inline .github/workflows/release.yaml:29]:
```suggestion
```
My bad. This line can be removed as we don't accept revision as input anymore

**Review by DomFleischmann — APPROVED** [review]

File: .github/workflows/publish.yaml
Comment (ca-scribner) [reply .github/workflows/publish.yaml]: For context, this line was added back when the upload action would exit with a failure code in any cases that it shouldn't push (like say if we just updated a PR that was in a git branched named like `something-other-than-branch/`).  Without this `if` statement, cases that **correctly should not publish anything** would show up in our github checks as an :x: and look like a failure.  

Are there any cases with the current action now which we want to avoid publishing?  I haven't tracked how the logic evolved.  My only concern about editing this is that we ensure the github CI summary shows an intuitive message, eg:
* :heavy_check_mark: if it published successfully
* :x: if it failed to do whatever was expected (build failure, etc)
* (something) if we *should not* publish **and** that we actually *do not* publish  
  * if this is not relevant in current action, we can ignore
  * If this is still a relevant case, I argue this should not give a :heavy_check_mark: status as seeing ":heavy_check_mark: Publish Charm" in the CI summary implies that we **published** something, when instead we've correctly **decided not to publish anything**
  * Similar argument against giving status of :x: - it implies something went wrong, rather than we correctly decided not to publish anything

Comment (ca-scribner) [reply .github/workflows/publish.yaml]: Regarding adding checks, I like adding checks that block publishing to main.  But what about publishing to branches?  I think there's good reason to publish to a branch even if linting/unit/integration tests fail - often we're debugging multiple things at once and it would be nice to have a branch available during that process.

File: .github/workflows/release.yaml
Comment (ca-scribner) [inline .github/workflows/release.yaml:1]: No need to separate it out now, but in future this feels like `release.yaml` could be a different PR as it does not depend on `publish.yaml`.  Would make it easier to approve and merge :)

**Review by ca-scribner — CHANGES_REQUESTED** [review]
See other comments.  Just adding this as a "request changes" as I think the others warrant a re-discuss before anyone merges this

**Review by DomFleischmann — CHANGES_REQUESTED** [review]
Lets discuss this a bit more.

File: .github/workflows/publish.yaml
Comment (ca-scribner) [inline .github/workflows/publish.yaml:71]: This duplicates everything from `ci.yaml`, so now maintainers need to write their testing twice (once in `ci.yaml`, once here).  I think I know what you're trying to do and I think we can achieve the goal without repeating the testing.

Comment (ca-scribner) [reply .github/workflows/publish.yaml:71]: Is the goal here to achieve:
* only publish to charmhub if CI passes (or my personal preference, only publish to `latest/edge` if CI passes, but always publish to `latest/edge/branchname` even if CI fails)
?

If yes, I think we can achieve this through a bit different methods without needing to repeat the tests

Comment (ca-scribner) [reply .github/workflows/publish.yaml:71]: I think there's at least two ways to achieve this.  Would appreciate other's feedback!

### Using workflow dispatch to reuse a workflow
See [reusing workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) for more detail, but effectively we can call an entire workflow similar to how we can call an action.  So we could do one of:
1. our `test.yaml` will call `publish.yaml` in cases where we should publish
2. our `publish.yaml` will call `test.yaml` before it runs

Either way I think it works.  It'll take a bit more effort to get the "publish to `latest/edge` IFF passing `test.yaml`, but always publish to `latest/edge/branch`", but there should be a way.

### Publishing to branch on `pull_request` events, publishing to `*/edge` on `push` events on selected branches.  

We could separate our publishing into two things:
```yaml
# publish-to-branch.yaml
on:
  pull_request:

#... publish
```

```yaml
# publish-to-edge.yaml
on:
  push:
    branches:
      - main
      - master
      - track/*

# workflow dispatch or otherwise call `test.yaml` to get tests

# publish, with `needs: above-tests`
```

I think this achieves all goals.

#### Variation on above using repo settings to protect against bad publishing

Instead of `publish-to-edge.yaml` above invoking a workflow dispatch, we could instead use repo settings to achieve the same thing.  `Settings->branchs->Branch protection rules` let us:
* `require "status checks" to pass before merging` anything into `main`/`track/*`:  this lets us require that all tests in `test.yaml` have passed before merging into a branch that gets published
* `require branches to be up to date before merging`: this lets us avoid "stale" mergers (where the CI would have failed because main has changed recently)

Comment (ca-scribner) [reply .github/workflows/publish.yaml:71]: From discussions with @sed-i, this is a possible way to implement the "using workflow dispatch" option.  We have workflow files of:

* `tests.yaml` <-- a workflow_dispatchable workflow for tests (lint, unit, ...)
* `publish.yaml` <-- a workflow_dispatchable workflow for publishing
* `on-pr.yaml` <-- on any pull_request event, this runs a workflow_dispatch for tests.yaml and publish.yaml (to branch), where publish runs no matter what
* `on-push.yaml` <-- on any push, this runs tests.yaml and then if tests passpublish.yaml (to edge)

File: .github/workflows/publish-to-branch.yaml
Comment (ca-scribner) [inline .github/workflows/publish-to-branch.yaml:8]: I think this can be removed, but you should confirm removing it will result in the behaviour you expect.  I think in the other kubeflow repos, we only have `branches` restrictions on `push` events, but not on `pull_request` events.  But maybe we hadn't considered it in the setting of the branch publishing, so I might be missing something

File: .github/workflows/publish-to-edge.yaml
Comment (ca-scribner) [inline .github/workflows/publish-to-edge.yaml:13]: I think to use the `ci.yaml` as a reusable workflow, `ci.yaml` must include an `on: workflow_dispatch` trigger event.  

If you haven't already, I suggest making a separate repo for testing how to use the workflow dispatch, etc., that way you can fine tune.  Unless testing by pushing to this PR is enough, if so then ignore :)

File: .github/workflows/publish-to-edge.yaml
Comment (ca-scribner) [inline .github/workflows/publish-to-edge.yaml:1]: I think we could rename this to `Test and publish to edge` or something like that.  

Also, at present, the testing will execute twice on push events (once in this workflow, and once because `ci.yaml` triggers on pushes).  I think we can also make `ci.yaml` only test on `pull_request` events, and then `push` event testing would be covered here.  Thoughts?

File: .github/workflows/publish-to-branch.yaml
Comment (ca-scribner) [inline .github/workflows/publish-to-branch.yaml:14]: We should confirm, but I think we're changing our philosophy here so that we **always** publish changes on github branchs to charmhub `latest/edge/branchname` branches, regardless of the github branch name (eg: we're no longer limiting "publish to branch" to only run on github branches named `branch/*`, but now we do it for all github branches)

Comment (sed-i) [reply .github/workflows/publish-to-branch.yaml:14]: (Passerby comment):

[Branch name](https://stackoverflow.com/a/3651867/3516684) would probably need to be sanitized:

```shell
$ git checkout -b some/crazy/branch.with.periods
Switched to a new branch 'some/crazy/branch.with.periods'

# HOWEVER

$ charmcraft release karma-k8s --channel=edge/some/crazy/branch.with.periods ...
Store operation failed:
- api-error: Invalid channel: edge/some/crazy/branch.with.periods
```

File: .github/workflows/publish-to-branch.yaml
Comment (natalian98) [reply .github/workflows/publish-to-branch.yaml:8]: I see that in many of our repos on `pull_request` events, so maybe we should keep it? For example, if I made a pr to your branch, I suppose it doesn't mean that the publish action should run

File: .github/workflows/publish-to-edge.yaml
Comment (natalian98) [reply .github/workflows/publish-to-edge.yaml:13]: That's right, I missed that point in `ci.yaml`. I think it should be `on: workflow_call` though - at least that seems to be working.
I made a [dummy repo](https://github.com/natalian98/template-operator/actions) to trigger the actions on pr and merge. The actions are obviously failing due to no secrets stored, but we can at least see that the tests workflow from `ci.yaml` is re-used and required by `publish-to-edge` as expected.

File: .github/workflows/publish-to-edge.yaml
Comment (natalian98) [reply .github/workflows/publish-to-edge.yaml]: I agree. `push` part removed from `ci.yaml`, this will be covered by the reusable worklfow

File: .github/workflows/publish-to-branch.yaml
Comment (ca-scribner) [reply .github/workflows/publish-to-branch.yaml]: Good point.  I think this would best be caught in the `upload-charm` action - does that sound right to you @sed-i?

Comment (sed-i) [reply .github/workflows/publish-to-branch.yaml]: I guess the answer is both:
- Own CI would be responsible for any custom name transformations.
- `upload-charm` would have to do the same generic sanitation for everything comes its way (or fail, depending on a flag passed?)

File: .github/workflows/publish-to-edge.yaml
Comment (simskij) [reply .github/workflows/publish-to-edge.yaml:13]: > workflow_call is used to indicate that a workflow can be called by another workflow.

> To manually trigger a workflow, use the workflow_dispatch event.

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows

File: .github/workflows/publish-to-branch.yaml
Comment (simskij) [reply .github/workflows/publish-to-branch.yaml:8]: I agree with @natalian98. PRs to feature branches should not lead to publishing.

File: .github/workflows/release.yaml
Comment (simskij) [inline .github/workflows/release.yaml:1]: In the observability team, we've opted for a somewhat different approach that I think is less likely to blow up due to erroneous user input. It also has the benefit of being more aligned with the idea of not skipping stages: 

https://github.com/canonical/grafana-agent-k8s-operator/blob/709f6a743db4c314e0760316c27578296b819d85/.github/workflows/promote.yaml#L14-L38

**Review by simskij — CHANGES_REQUESTED** [review]

**Comment by natalian98** [comment]
Closing as we reached an agreement internally.

**Pull request closed without merging**
